nemo/collections/asr/models/aed_multitask_models.py

```diff
@@ -95,83 +96,81 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
-        )
-        self.log_softmax.mlp.layer0.weight = self.transf_decoder.embedding.token_embedding.weight
+        # Auto inject vocab size for `get_transformer`
+        with open_dict(transf_decoder_cfg_dict):
+            if 'config_dict' in transf_decoder_cfg_dict:
+                transf_decoder_cfg_dict['config_dict']['vocab_size'] = vocab_size
+
+        self.transf_decoder = EncDecMultiTaskModel.from_config_dict(transf_decoder_cfg_dict)
+
+        # Setup token classifier
+        with open_dict(self.cfg.head):
+            self.cfg.head.num_classes = vocab_size
+
+        self.log_softmax = EncDecMultiTaskModel.from_config_dict(self.cfg.head)
+        if isinstance(self.log_softmax, TokenClassifier):
+            self.log_softmax.mlp.layer0.weight = self.transf_decoder.embedding.token_embedding.weight
+        else:
+            raise ValueError(f"Expected TokenClassifier head, got {type(self.log_softmax)}")
         std_init_range = 1 / self.transf_decoder.hidden_size ** 0.5
         self.transf_decoder.apply(lambda module: transformer_weights_init(module, std_init_range))
         self.log_softmax.apply(lambda module: transformer_weights_init(module, std_init_range))
 
-        # Beam Search decoding
-        self.beam_search = BeamSearchSequenceGenerator(
-            embedding=self.transf_decoder.embedding,
-            decoder=self.transf_decoder.decoder,
-            log_softmax=self.log_softmax,
-            max_sequence_length=self.transf_decoder.max_sequence_length,
-            beam_size=self.cfg.beam_search.beam_size,
-            bos=self.tokenizer.bos_id,
-            pad=self.tokenizer.pad_id,
-            eos=self.tokenizer.eos_id,
-            len_pen=self.cfg.beam_search.len_pen,
-            max_delta_length=self.cfg.beam_search.max_generation_delta,
+        # Setup decoding objects
+        decoding_cfg = self.cfg.get('decoding', None)
+
+        # In case decoding config not found, use default config
+        if decoding_cfg is None:
+            decoding_cfg = OmegaConf.structured(MultiTaskDecodingConfig)
+            with open_dict(self.cfg):
+                self.cfg.decoding = decoding_cfg
+
+        self.decoding = MultiTaskDecoding(
+            decoding_cfg=self.cfg.decoding,
+            transformer_decoder=self.transf_decoder,
+            log_softmax_module=self.log_softmax,
+            tokenizer=self.tokenizer,
         )
+
         # TO DO: remove this hardcoded context size for AR decoding
-        self.context_len_for_AR_decoding = 5
+        self.context_len_for_AR_decoding = self.cfg.get("context_len_for_AR_decoding", 5)
 
         # Define autoregressive CE loss
-        self.transf_loss = SmoothedCrossEntropyLoss(
-            pad_id=self.tokenizer.pad_id, label_smoothing=self.cfg.label_smoothing
-        )
+        with open_dict(self.cfg.loss):
+            self.cfg.loss.pad_id = self.tokenizer.pad_id
+
+        self.loss = EncDecMultiTaskModel.from_config_dict(self.cfg.loss)
```

Comment:
any reason why we removed `label_smoothing` here ? yaml config has `label_smoothing` parameter.

Reply:
Forgot. Changed the config to read it from the top of the config with

```yaml
  loss:
    _target_: nemo.collections.common.losses.smoothed_cross_entropy.SmoothedCrossEntropyLoss
    label_smoothing: ${model.label_smoothing}
    pad_id: null
```